jane/doc/extensions/uniqueness/pitfalls.md

```diff
@@ -0,0 +1,115 @@
+# Some Pitfalls of Uniqueness
+
+This document outlines some common pitfalls that may come up when trying out
+uniqueness, as well as some suggested workarounds. Over time, this list may grow
+(as experience discovers new things that go wrong) or shrink (as we deploy new
+compiler versions that ameliorate some issues).
+
+If you want an introduction to uniqueness, see the [introduction](intro.md).
+
+## Annotating unique return values
+
+When the compiler infers uniqueness, it will only mark values as `unique` if
+they are actually used as `unique`. For example, if create a smart constructor:
+
+```ocaml
+let mk () = { x = 1 }
+```
+
+This value is not unique by default. This helps the compiler identify static
+allocations: as long as `mk ()` returns an `aliased` value, the compiler can
+lift out the allocation so that `mk ()` returns the same pointer on each
+invokation. However, you might want to annotate this allocation as `unique`:
+
+```ocaml
+let mk () = unique_ { x = 1 }
+```
+
+Unfortunately, this will not work: while the allocation is now indeed `unique`,
+the compiler immediately casts it to `aliased` so that `mk` still returns an
+`aliased` value. The only way to reliably change the return mode of `mk` to be
+`unique` is through a type signature:
+
+```ocaml
+let mk : unit -> t @ unique = fun () -> { x = 1 }
+```
+
+Most times, it will not be necessary to manually annotate a return value as
+`unique`. While the compiler defaults the return mode to `aliased`, it will make
+the return mode `unique` if there is any unique use of the result of `mk ()`:
+
+```ocaml
+let use () = free (mk ())
+```
+
+This code checks even if `mk` is not annotated, because, now that the need as
+arisen, the compiler has inferred the return mode of `mk` to be `unique`.
+However, this analysis happens on a per module basis. If we abstract `mk` into
+its own module we will get a mode error:
+
+```ocaml
+module Mk = struct
+  let mk () = { x = 1 }
+end
+
+let use () = free (Mk.mk ())
+                  ^^^^^^^^^^
+Error: This value is "aliased" but expected to be "unique".
+```
+
+When abstracting functions that return unique values into a module, we need to
+provide an explicit mode annotation in the signature.
+
+## Threading unique values
+
+While unique values allow safe mutation, they require a coding style that is
```

Comment:
These docs don't include unique mutation / strong updates - is that supported yet? If so it should be covered in the intro

Reply:
Overwrite is not yet supported -- not even as an alpha language extension.